### ISSUE TITLE
Investigate potential rounding error in getMaximalNoSwapLPAmount

### ISSUE THREAD

**Issue by AlpinYukseloglu** [osmosis-labs/osmosis#1908]
## Background

`getMaximalNoSwapLPAmount` uses QuoInt to calculate the share ratio needed to enter the pool, which does floor division and therefore might underestimate the amount of input liquidity needed from each token. Any underestimate related to this might be eclipsed by the subsequent round-up for each token's liquidity, but it's not clear to me whether this is always the case.

https://github.com/osmosis-labs/osmosis/blob/85529c76683f2d2f2ffb2121125baf0e39463f84/x/gamm/keeper/pool_service.go#L232

## Suggested Design

- If this is an issue, round up on the share ratio instead
- If not, clarify the rounding logic in comments

## Acceptance Criteria

N/A

**Comment by p0mvn** [comment]
Do you propose to convert `totalSharesAmount` to `Dec`, perform the division, and round up?

I think that makes sense to me

**Comment by p0mvn** [comment]
I also see no unit tests for `getMaximalNoSwapLPAmount`. I think we should invest in unit testing this function with calculations verified by Wolfram.

**Comment by p0mvn** [comment]
Another architectural issue is that we are passing a `PoolI` to this function just to perform some calculations by using the pool:
```
getMaximalNoSwapLPAmount(ctx sdk.Context, pool types.PoolI, shareOutAmount sdk.Int)
```

where it is called like:
```
neededLpLiquidity, err := getMaximalNoSwapLPAmount(ctx, pool, shareOutAmount)
```

Instead, I think this function should be defined on the pool interface so that we can call it in the following way:
```
neededLpLiquidity, err := pool.GetMaximalNoSwapLPAmount(ctx, shareOutAmount)
```

**Comment by AlpinYukseloglu** [comment]
> Do you propose to convert `totalSharesAmount` to `Dec`, perform the division, and round up?
> 
> I think that makes sense to me

Yes, my suggestion is to make the following change:
https://github.com/osmosis-labs/osmosis/blob/85529c76683f2d2f2ffb2121125baf0e39463f84/x/gamm/keeper/pool_service.go#L232
to this:

`shareRatio := shareOutAmount.ToDec().QuoInt(totalSharesAmount)`

cc: @ValarDragon, would love to validate whether this is a valid concern before pushing a PR for this change



> I also see no unit tests for `getMaximalNoSwapLPAmount`. I think we should invest in unit testing this function with calculations verified by Wolfram.

I agree, will make a separate testing-related issue for this

**Comment by AlpinYukseloglu** [comment]
cc: @ValarDragon

**Comment by ThanhNhann** [comment]
Hi @p0mvn, I see that the unit test for `getMaximalNoSwapLPAmount ` still doesn't have, Can I do it? And about 

> Instead, I think this function should be defined on the pool interface so that we can call it in the following way:
And how about this?

**Comment by p0mvn** [comment]
Hi @ThanhNhann 

Yes, that would be great! Please feel free to pick up the tests work. However, do you mind creating a separate issue for it and assigning yourself to that?

The original problem here seems to be something else that is still valid.

> > Instead, I think this function should be defined on the pool interface so that we can call it in the following way:

> And how about this?

I think that still makes sense. However, I would define it as the `TraditionalAmmInterface` instead of `PoolI`. I think we should make a separate issue for this as well. Would you be interested in spearheading that?

**Comment by ThanhNhann** [comment]
> Yes, that would be great! Please feel free to pick up the tests work. However, do you mind creating a separate issue for it and assigning yourself to that?


Yeah sure, I will create a issue for this, but can't assign myself because I don't really added in this repo, can you add me to the repo sir @p0mvn 

> I think that still makes sense. However, I would define it as the TraditionalAmmInterface instead of PoolI. I think we should make a separate issue for this as well. Would you be interested in spearheading that?

you mean a new interface `TraditionalAmmInterface ` right? I don't see the name in code, if true, I will have some questions before work

**Comment by p0mvn** [comment]
I don't have access but I pinged ValarDragon to add you to the repo. 

> you mean a new interface TraditionalAmmInterface in the file below right?

yeah

**Comment by ThanhNhann** [comment]
Thank sir @p0mvn, working on issues
